src/key.rs

```diff
@@ -1718,6 +1718,31 @@ mod test {
             "0100000000000000020000000000000003000000000000000400000000000000"
         );
     }
+    #[test]
+    fn test_display_obfuscation() {
+        #[rustfmt::skip]
+        static SK_BYTES: [u8; 32] = [
+            0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+            0xff, 0xff, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00,
+            0x63, 0x63, 0x63, 0x63, 0x63, 0x63, 0x63, 0x63,
+        ];
+
+        let sk = SecretKey::from_slice(&SK_BYTES).expect("sk");
+
+        #[cfg(feature = "std")]
+        {
+            let got = format!("{:?}", sk);
+            let want = "SecretKey(#2f4d520bd6874292)";
```

Comment:
I've just realized, I don't believe this value is stable - different compiler version/OSes/... may produce something different. We need to check these properties instead:

```rust
got.len() == im_too_lazy_to_count_it;
got.starts_with("SecretKey(#");
got.ends_with(")");
got[11..(got.len() - 1)].chars().all(|c| c.is_ascii_hexdigit());
!sk.display_secret().to_string().contains(&got[11..(got.len() - 1)]);
```

Reply:
Unrelated to this PR :)

Reply:
It **is** related. You've added a clearly broken test and the only reason it didn't manifest as broken is that nobody changed the hash algo between 1.63 and 1.80. [The doc](https://doc.rust-lang.org/stable/std/collections/hash_map/struct.DefaultHasher.html) specifically says: "The internal algorithm is not specified, and so it and its hashes should not be relied upon over releases."

If you don't want to property-test it then please at least remove it. (with a comment that's why) Also it might be prudent to use `RandomState::new().build_hasher()` at least in tests to make sure nobody adds the test back by accident but potentially to also frustrate an attacker that would like to somehow recover some bits of the private key. (Does it make sense @apoelstra ?)

Reply:
oh woops I can't read diffs, I was thinking that because I did not touch the `std` behaviour I did not write the test. My bad. I dropped the test, I think we are all convinced that this PR does what it says it does, re truncation. (I was very testy [0] yesterday afternoon.)

[0] definition of 'testy': https://duckduckgo.com/?q=testy+def&t=canonical&ia=web

Reply:
@Kixunil I think, ultimately Rust's reduced-siphash would have to be super broken for it to actually leak any bits of the key, and in the limit it can only leak 64 bits because that's how big its output is, but it *does* make me a little nervous to treat it as a random oracle like this.

Using a randomized hasher would help with this, at the cost of making the "key fingerprints" no longer work as fingerprints when they occur in logs. I think that would be super frustrating and likely even confusing, so we shouldn't do it in non-test contexts. (And yes, I appreciate that users already shouldn't be expecting these "fingerprints" to stay consistent because stdlib might change them....but in practice they probably will.)

Honestly I'm inclined to get rid of the `Hasher` stuff entirely and just put a `<secret key; please enable `hashes` in rust-secp to see a fingerprint>` type message in its place, just as we do for nostd users. I definitely wouldn't have written code that uses siphash like this and it looks like it originates in a time when I was much less strict in reviewed code than I would be with my own.

Reply:
But will ACK/merge this since it's a strict improvement on the status quo.

Reply:
All good points, just note that the output of `Debug` is normally not considered stable.

> just put a `<secret key; please enable `hashes` in rust-secp to see a fingerprint>` type message in its place

That sounds great to me! Let's do it.